apps/omg_watcher_info/test/support/factory.ex

```diff
@@ -58,12 +58,69 @@ defmodule OMG.WatcherInfo.Factory do
 
   alias OMG.WatcherInfo.DB
 
+  @eth OMG.Eth.RootChain.eth_pseudo_address()
+
+  @doc """
+  Block factory.
+
+  Generates a block in an incremental blknum of 1000, 2000, 3000, etc.
+  """
   def block_factory() do
     %DB.Block{
       blknum: sequence(:block_blknum, fn seq -> seq * 1000 end),
-      hash: sequence(:block_hash, fn seq -> <<seq::256>> end),
+      hash: insecure_random_bytes(32),
       eth_height: sequence(:block_eth_height, fn seq -> seq end),
       timestamp: sequence(:block_timestamp, fn seq -> seq * 1_000_000 end)
     }
   end
+
+  @doc """
+  Transaction factory.
+
+  Generates a transaction without any transaction output and no associated block.
+
+  To generate a transaction with closest data to production, consider associating the transaction
+  to a block and generating transaction outputs associated with this transaction.
+  """
+  def transaction_factory() do
+    %DB.Transaction{
+      txhash: sequence(:transaction_hash, fn seq -> <<seq::256>> end),
+      txindex: 0,
+      txbytes: insecure_random_bytes(32),
+      sent_at: DateTime.utc_now(),
+      metadata: insecure_random_bytes(32),
+      block: nil,
+      inputs: [],
+      outputs: []
+    }
+  end
+
+  @doc """
+  Txoutput factory.
+
+  Generates a txoutput. The txindex, oindex and spending_tx_oindex are defaulted to 0.
+  These two values need to be overridden depending on the transaction you aim to build.
+  """
+  def txoutput_factory() do
+    owner = OMG.TestHelper.generate_entity()
+
+    %DB.TxOutput{
+      blknum: insert(:block).blknum,
+      txindex: 0,
+      oindex: 0,
+      owner: owner.addr,
+      amount: 100,
+      currency: @eth,
+      proof: insecure_random_bytes(32),
+      spending_tx_oindex: 0,
+      child_chain_utxohash: insecure_random_bytes(32),
+      creating_transaction: insert(:transaction),
+      spending_transaction: nil
+    }
+  end
```

Comment:
Generally, do we want the default test objects that the factories create to be:
1. consistent and valid with our business logic (meaning parent and/or children objects block <-> transaction <-> utxo may need to be created together via `build()` calls in the factories or using `params_with_assocs()`
2. or at a minimum the object can be inserted into the DB and not violate DB constraints although it might be business rules may be violated.

I think 1) seems like the most reasonable thing to grow the factories to produce as defaults. But only over time as more tests are moved over.

2) database insertable default object seems like a good baseline to start with though.

Just some things that caught my eye, if you're interested in putting them this PR...

1. The UTXO being created in this factory gets its `blknum` from `blknum: insert(:block).blknum`. But will that be the same blknum as the transaction that created it? `creating_transaction: insert(:transaction)` is called, but the default transaction factory set its `block` to nil.
2. Does `insert(:transaction)` work ok in this PR given the DB key constraints, but the nil transaction block?
3. Correct sequencing of `blknum, txindex, oindex`. Eg: if we want to create a deposit then is_deposit(blknum, txindex, oindex)` should return true. see [code](https://github.com/omisego/elixir-omg/blob/master/apps/omg/lib/omg/utxo/position.ex#L84-L86)

Reply:
☝️ I'm thinking about this stuff in my PR too.

Reply:
One last thing I forgot to ask... why the use of `insert()` in the factory instead of `build()`.  ExMachina docs discourage that a bit.

To have the `block` and `transaction` objects built and inserted in the txoutput factory can you use `params_with_assocs(:txoutput)`. That worked well for me.

Reply:
On your first comment, I was thinking the same too but chose the 2nd approach because different use cases demand data to be constructed in different ways. E.g. a spent utxo will have both `creating_txhash` and `spending_txhash`, then you'll need to construct two transactions. While an unspent utxo has only `creating_txhash` and only one transaction is needed, etc. etc.

So in this case I chose to let each tests determine its own needs while the factories take care of inserted defaults that the tests can chose not to care about.

If these primitive factories are built with in "consistent, valid with our business logic" then we end up with highly-coupling fixtures again.

I'm supportive of building more derived factories on top to fit certain use cases though.

Reply:
> ExMachina docs discourage that a bit.

Can you point me to that part? I'm having trouble finding it.

Reply:
https://hexdocs.pm/ex_machina/readme.html#ecto-associations
> Using insert/2 in factory definitions may lead to performance issues and bugs, as records will be saved unnecessarily.

In terms of your idea:
> So in this case I chose to let each tests determine its own needs while the factories take care of inserted defaults that the tests can chose not to care about.

I was thinking about this as well and was able to get along just fine using only `build()` and `params_with_assocs(:txoutput)` which lets us build a txoutput and not care about the parent `transaction` about and the `transaction`'s parent object `block`. I does require taking full control of the factory though to make sure the blknum of the txoutput and the blknum of the `creating_transaction` are the same.

If I want to specifically override the `txoutput`'s `transaction` and/or `block`, using attr overrides can be used. Eg:

`build(:txoutput,: creating_transaction: my_creating_txn)` 
or using pipes `build(:txoutput) |> with_creating_txn(...) |> with_spending_txn(...)`

There is also the question with the code, specifically:
```
    %DB.TxOutput{
      blknum: insert(:block).blknum,
      <snip>
      creating_transaction: insert(:transaction),
      <snip>
    }
```

From a business logic perspective the `creating_transaction` above should have a `blknum` that is equal to the txoutput's `blknum`. But building it this way, it's not clear that that is true?

Reply:
How about https://github.com/omisego/elixir-omg/pull/1224/commits/3a450bf94f08c3a4e8985fe37e52d12346733624 now? The txoutput_factory should now have consistent block and transaction.